src/edu/psu/sweng500/team8/play/GameSession.java

```diff
@@ -55,24 +53,61 @@ public Board getGameBoard() {
 	 * If we want to use return object, we maybe able to use Enum to show previously defined exception cases.
 	 * 
 	 *
-	 * @param cellCoordinates
+	 * @param cell
 	 * @param number
 	 */
-	public void enterNumber(CellCoordinates cellCoordinates, int number) {
-
+	public void enterNumber(Cell currentCell, int number) {
+		
 		/* keep track of the last action*/
 		SudokuAction sudokuAction = new SudokuAction(new CellGrid(board.getCellGrid()));
 		
-		/* updating number */
-		Cell currentCell = board.getCell(cellCoordinates.getRowIndex(), cellCoordinates.getColumnIndex());
-		currentCell.setNumber(number);
+		boolean isDelete = false;
 		
-		actionManager.addAction(sudokuAction);
+		if (number == 0) {
+			isDelete = true;			
+		}
+		
+		if (isDelete) {
+			currentCell.clearNumber();
+		} else {
+			currentCell.setNumber(number);	
+		}
+		
+		/* FIXME - PencilMarkManager updating/wiping out PencilMark number matching entered number */
+		if (!isDelete) {
+			for (int i = 0; i < 9; i++) {
+				for (int j = 0; j < 9; j++) {
+					Cell eachCell = this.getGameBoard().getCellGrid().getCell(i, j);
+					enterPencilMark(eachCell, number, false);
+				}
+			}			
+		}
 		
-		/* TODO - PencilMarkManager updating/wiping out PencilMark number matching entered number */
+		actionManager.addAction(sudokuAction);
 	}
+
 	
 	/**
+	 * TODO implement
+	 * @param currentCell
+	 * @param number
+	 */
+	public void enterPencilMark(Cell currentCell, int number, boolean isEnter) {
+		
+		/* keep track of the last action*/
+		SudokuAction sudokuAction = new SudokuAction(new CellGrid(board.getCellGrid()));

```

Comment:
We can limit the number of items that can be reverted back. A simple example is MS Windows's Notepad. You can only do so many UNDOs. I would say 10 actions (arbitrary).
All we are doing is storing Integers 81*32bit+81*81*32*bit =  6kB. Even if we store 10, it's only 60kB. (of course, it's bigger than that as we store some other pointers, but I think you see my point). Since we only use 1-9, we can use short or even byte to store our numbers. I don't think it really matters as we are not dealing with a machine that needs to crunch a couple MBs.